### PR TITLE
Add Cypress test to patient file upload

### LINKED ISSUE #7086: Add Cypress Test for Patient file upload
**Description Of the Test**

Add a new cypress test to the patient file upload on the patient detail page, and cover the following flow in the test file

- [ ] Upload a new file to the patient detail page. (Both Audio and file)
- [ ] Edit the file name of an existing uploaded file.
- [ ] Archive the existing file and verify its move to the archive section along with archive details.
- [ ] Verify the uploaded file can only be modified by the author, district admin, and above users.
- [ ] Verify file download is possible for all users.

**Note**

- add query parameter, notification, and API verification wherever necessary
- follow the POM approach while coding
- create a new file named "patient_detailpage.cy.ts" and page object file "PatientFileupload.ts" 

**Screenshot**

image

### PR DESCRIPTION
**Pull request by AshrafMd-1**
## Proposed Changes

- Fixes #7086 
- Add IDs to tags
- Add cypress testing

@coronasafe/care-fe-code-reviewers @coronasafe/code-reviewers

## Merge Checklist

- [ ] Add specs that demonstrate bug / test a new feature.
- [ ] Update [product documentation](https://docs.coronasafe.network/coronasafe-care-documentation/architecture/architecture-and-layering-of-care).
- [ ] Ensure that UI text is kept in I18n files.
- [ ] Prep screenshot or demo video for changelog entry, and attach it to issue.
- [ ] Request for Peer Reviews
- [ ] Completion of QA

### CODE REVIEW

**Comment by AshrafMd-1** [comment]
@nihal467 for the following 

[ ] Verify the uploaded file can only be modified by the author, district admin, and above users.
[ ]  Verify file download is possible for all users.

Do I need to sign to each kind of user and check or is there a different way?

cc: @rithviknishad

**Comment by nihal467** [comment]
@AshrafMd-1 

way of approach should be like , use two nurse type and one district admin

Test - 1 Part-1 : Verify the uploaded file can only be modified by the author, district admin, and above users.
- [ ] First, upload a file to a patient file using the nurse user ID-1, and edit the file successfully.
- [ ] Now, log out as nurse user ID-1 and log in as nurse user ID-2. Then try to edit the same file. Expected behavior: "Nurse user ID-2 shouldn't be able to edit the file uploaded by nurse user ID-1." Verify this behavior.
- [ ] Now, log out as nurse user ID-2 and log in as a district admin. Edit the file uploaded by nurse user ID-1. Expected behavior: "The admin should be able to edit the file uploaded by nurse user ID-1."

Test - 1 Part-2 : Verify file download is possible for all users.

- [ ] Verify that the download functionality is working for nurse user ID's and the district admin, along with the part-1 test.

@AshrafMd-1 Answer to the question : Only check for district admin and nurse userID

**Comment by AshrafMd-1** [comment]
@nihal467 done I understood

Also other small doubt what patient should I use

Like I created a patient but it was not coming on cypress

**Comment by nihal467** [comment]
@AshrafMd-1 don't create a new patient, just use a patient in the existing dummy data

**Comment by AshrafMd-1** [comment]
@nihal467 is there any credentials for 
nurse id-1 and nurse id-2 or should I create a new user first

**Comment by AshrafMd-1** [comment]
@nihal467 can you check this once 

cc: @rithviknishad

File: cypress/e2e/patient_spec/patient_detailpage.cy.ts [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+    cy.clearLocalStorage(/filters--.+/);
+    cy.awaitUrl("/patients");
+  });
+
+  it("Record an audio and save it", () => {
+    patientPage.visitPatient("Dummy Patient 3");
+    patientFileUploadPage.visitPatientDetailsPage();
+    patientFileUploadPage.recordAudio();
+    patientFileUploadPage.clickUploadAudioFile();
+    patientFileUploadPage.verifySuccessNotification(
```
Comment (nihal467) [inline cypress/e2e/patient_spec/patient_detailpage.cy.ts]:
```suggestion
    cy.verifyNotification("File Uploaded Successfully");
```

File: cypress/e2e/patient_spec/patient_detailpage.cy.ts [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+      "File Uploaded Successfully"
+    );
+  });
+
+  it("Upload a file", () => {
+    patientPage.visitPatient("Dummy Patient 4");
+    patientFileUploadPage.visitPatientDetailsPage();
+    patientFileUploadPage.uploadFile();
+    patientFileUploadPage.clickUploadFile();
+    patientFileUploadPage.verifySuccessNotification(
```
Comment (nihal467) [inline cypress/e2e/patient_spec/patient_detailpage.cy.ts]:
```suggestion
    cy.verifyNotification("File Uploaded Successfully");
```

File: cypress/e2e/patient_spec/patient_detailpage.cy.ts [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+
+  it("Edit file name", () => {
+    patientPage.visitPatient("Dummy Patient 4");
+    patientFileUploadPage.visitPatientDetailsPage();
+    patientFileUploadPage.verifyFileEditOption(true);
+    patientFileUploadPage.editFileName(
+      `Cypress File ${new Date().getTime().toString().slice(9)}`
+    );
+    patientFileUploadPage.clickSaveFileName();
+    patientFileUploadPage.verifySuccessNotification(
```
Comment (nihal467) [inline cypress/e2e/patient_spec/patient_detailpage.cy.ts]:
```suggestion
    cy.verifyNotification("File name changed successfully");
```

File: cypress/e2e/patient_spec/patient_detailpage.cy.ts [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+      "File name changed successfully"
+    );
+  });
+
+  it("Archive file and verify it", () => {
+    patientPage.visitPatient("Dummy Patient 4");
+    patientFileUploadPage.visitPatientDetailsPage();
+    patientFileUploadPage.archiveFile();
+    patientFileUploadPage.clickSaveArchiveFile();
+    patientFileUploadPage.verifySuccessNotification(
```
Comment (nihal467) [inline cypress/e2e/patient_spec/patient_detailpage.cy.ts]: cy.verifyNotification("File archived successfully");

File: cypress/e2e/patient_spec/patient_detailpage.cy.ts [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+    patientFileUploadPage.verifyArchiveFile();
+  });
+
+  it("Verify the uploaded file can be edited by author", () => {
+    loginPage.login("dummynurse1", "Coronasafe@123");
+    patientPage.visitPatient("Dummy Patient 5");
+    patientFileUploadPage.visitPatientDetailsPage();
+    patientFileUploadPage.uploadFile();
+    patientFileUploadPage.clickUploadFile();
+    patientFileUploadPage.verifySuccessNotification(
```
Comment (nihal467) [inline cypress/e2e/patient_spec/patient_detailpage.cy.ts]: replace with cy.verifyNotification

File: cypress/e2e/patient_spec/patient_detailpage.cy.ts [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+      "File Uploaded Successfully"
+    );
+    patientFileUploadPage.verifyFileEditOption(true);
+    patientFileUploadPage.editFileName(
+      `Cypress File ${new Date().getTime().toString().slice(9)}`
+    );
+    patientFileUploadPage.clickSaveFileName();
+    patientFileUploadPage.verifySuccessNotification(
```
Comment (nihal467) [inline cypress/e2e/patient_spec/patient_detailpage.cy.ts]: replace with cy.verifyNotification

File: cypress/e2e/patient_spec/patient_detailpage.cy.ts [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+    loginPage.loginAsDisctrictAdmin();
+    patientPage.visitPatient("Dummy Patient 5");
+    patientFileUploadPage.visitPatientDetailsPage();
+    patientFileUploadPage.verifyFileEditOption(true);
+    patientFileUploadPage.editFileName(
+      `Cypress File ${new Date().getTime().toString().slice(9)}`
+    );
+    patientFileUploadPage.clickSaveFileName();
+    patientFileUploadPage.verifySuccessNotification(
+      "File name changed successfully"
```
Comment (nihal467) [inline cypress/e2e/patient_spec/patient_detailpage.cy.ts]: replace with cy.verifyNotification

File: cypress/pageobject/Patient/PatientFileupload.ts [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+  }
+
+  editFileName(newFileName: string) {
+    cy.get("#edit-file-name").click().scrollIntoView();
+    cy.get("#editFileName").clear().type(newFileName);
+  }
+
+  clickSaveFileName() {
+    cy.intercept("PATCH", "**/api/v1/files/**").as("saveFileName");
+    cy.get("#submit").click();
```
Comment (nihal467) [inline cypress/pageobject/Patient/PatientFileupload.ts]: cy.submitButton("text inside the button");

File: cypress/pageobject/Patient/PatientFileupload.ts [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+    cy.get("#file-name").then(($el: string) => {
+      fileName = $el.text().split(":")[1].trim();
+    });
+    cy.get("#archive-file").click().scrollIntoView();
+    cy.get("#editFileName").clear().type("Cypress File Archive");
+  }
+
+  clickSaveArchiveFile() {
+    cy.intercept("PATCH", "**/api/v1/files/**").as("saveArchiveFile");
+    cy.get("#submit").click();
```
Comment (nihal467) [inline cypress/pageobject/Patient/PatientFileupload.ts]: cy.submitButton("text inside the button");

File: cypress/pageobject/Patient/PatientFileupload.ts [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+    cy.get("#preview-file").should(status ? "be.visible" : "not.be.visible");
+  }
+
+  downloadFile() {
+    cy.intercept("GET", "**/api/v1/files/**").as("downloadFile");
+    cy.get("#preview-file").click();
+    cy.wait("@downloadFile").its("response.statusCode").should("eq", 200);
+  }
+
+  verifySuccessNotification(msg: string) {
```
Comment (nihal467) [inline cypress/pageobject/Patient/PatientFileupload.ts]: delete this variable since we will be using the verify notification from command.ts file

**Review by nihal467 — CHANGES_REQUESTED** [review]

File: cypress/e2e/patient_spec/patient_detailpage.cy.ts [hunk 11]
```diff
@@ -0,0 +1,133 @@
...
+  });
+
+  it("Record an audio and save it", () => {
+    patientPage.visitPatient("Dummy Patient 3");
+    patientFileUploadPage.visitPatientDetailsPage();
+    patientFileUploadPage.recordAudio();
+    patientFileUploadPage.clickUploadAudioFile();
+    patientFileUploadPage.verifySuccessNotification(
+      "File Uploaded Successfully"
+    );
```
Comment (nihal467) [inline cypress/e2e/patient_spec/patient_detailpage.cy.ts]: along with the notification, verify it is visible in the file list

File: cypress/e2e/patient_spec/patient_detailpage.cy.ts [hunk 12]
```diff
@@ -0,0 +1,133 @@
...
+  });
+
+  it("Upload a file", () => {
+    patientPage.visitPatient("Dummy Patient 4");
+    patientFileUploadPage.visitPatientDetailsPage();
+    patientFileUploadPage.uploadFile();
+    patientFileUploadPage.clickUploadFile();
+    patientFileUploadPage.verifySuccessNotification(
+      "File Uploaded Successfully"
+    );
```
Comment (nihal467) [inline cypress/e2e/patient_spec/patient_detailpage.cy.ts]: along with the notification, verify it is visible in the file list

File: cypress/e2e/patient_spec/patient_detailpage.cy.ts [hunk 13]
```diff
@@ -0,0 +1,133 @@
...
+    patientPage.visitPatient("Dummy Patient 4");
+    patientFileUploadPage.visitPatientDetailsPage();
+    patientFileUploadPage.verifyFileEditOption(true);
+    patientFileUploadPage.editFileName(
+      `Cypress File ${new Date().getTime().toString().slice(9)}`
+    );
+    patientFileUploadPage.clickSaveFileName();
+    patientFileUploadPage.verifySuccessNotification(
+      "File name changed successfully"
+    );
```
Comment (nihal467) [inline cypress/e2e/patient_spec/patient_detailpage.cy.ts]: along with the notification, verify it is visible in the file list

File: cypress/pageobject/Patient/PatientFileupload.ts [hunk 14]
```diff
@@ -0,0 +1,90 @@
...
+  verifyArchiveFile() {
+    cy.get("#archived-files").click();
+    cy.get("#file-name").then(($el) => {
+      const text = $el.text().split(":")[1].trim();
+      cy.expect(text).to.eq(fileName);
+    });
+  }
+
+  verifyFileEditOption(status: boolean) {
+    cy.get("#file-div").should(
```
Comment (nihal467) [inline cypress/pageobject/Patient/PatientFileupload.ts]: verify the archive reason by clicking "more details"

**Comment by nihal467** [comment]
@AshrafMd-1 

Test 1: Record an Audio and download the file.

- Record an audio.
- Verify its success through notification and reflection in the list.
- Verify the file and ensure its reflection in the list as well as notification.
- Try to download the file and verify its success using API response with cy.intercept.

By performing the above tasks, you can remove the following tests:

"Verify uploaded file can be edited by author"
"Verify that file download is possible for district admin and above"
"Verify that file download is possible for author"

Test 2: Upload a File and archive it

- Upload a file and verify its reflection in the list.
- Archive the file and verify the file list along with the reason.

By performing the above tasks, you can remove the test "Archive file and verify it."

Test 3: User-level Based Permission for File Modification

- Upload a file using Nurse ID-1 and download it.
- Edit the file using Nurse ID-1.
- Log in as Nurse ID-2 and verify that the file can't be edited.
- Log in as district admin and verify that the file can be edited.

You can remove the test "Verify that file download is possible for users below district admin."

so by combining tests, it should achieve the same result by saving much more time

**Comment by AshrafMd-1** [comment]
@nihal467 
for 
> Test 1: Record an Audio and download the file.
> - Try to download the file and verify its success using API response with cy.intercept.

Its a anchor tag so verifying with API is not possible so what if I change the tests in this format 

image


how about this 

> test1 : Upload a file and download file

>test2 : record a audio and archive it

**Comment by AshrafMd-1** [comment]
@nihal467 can u check once

**Comment by AshrafMd-1** [comment]
@nihal467 can you check once why Cypress is failing 
because it works perfectly on my computer 

I feel it's due to not having the nurse logins

**Comment by AshrafMd-1** [comment]
@rithviknishad  can u remove the merge-conflict label 

also is it possible to add nurse login to dummy database in care ?

**Comment by rithviknishad** [comment]
Yup, you can load the dummy data on to a fresh volume, add the required users and dump it again as json

**Comment by nihal467** [comment]
> @nihal467 for
> 
> > Test 1: Record an Audio and download the file.
> > 
> > * Try to download the file and verify its success using API response with cy.intercept.
> 
> Its a anchor tag so verifying with API is not possible so what if I change the tests in this format
> 
> image
> how about this
> 
> > test1 : Upload a file and download file
> 
> > test2 : record a audio and archive it

@AshrafMd-1 when you upload a file the API "https://care.coronasafe.in/api/v1/files/" gives you 201 responses verify that using the cy.intercept .

_Note: Its fine to only do api verification for file upload in this case and no need of API verification on downloads_

**Comment by AshrafMd-1** [comment]
@nihal467 I have made changes as you have referenced 
Can you check once

**Comment by nihal467** [comment]
@AshrafMd-1 your test is failing

**Comment by AshrafMd-1** [comment]
@nihal467 once this PR gets merged in care related to nurse logins, then I will rerun the tests 

https://github.com/coronasafe/care/pull/1985

**Comment by nihal467** [comment]
@AshrafMd-1 the test is again failing, the backend is merged

**Comment by AshrafMd-1** [comment]
@nihal467 I am not able to understand the error 
The tests works locally but when I upload here no patients are found 

- I did check the district `ernakulam`
- The home facility `dummy facility 40`

But then the screenshots says no patients are found 

So can you give me some guidance on what to do?

**Comment by nihal467** [comment]
image

@AshrafMd-1 check the login id user access, whether it is having the particular facility access

**Comment by AshrafMd-1** [comment]
@nihal467 

Here is the nurse1 profile
image

Here is its data

image


But its showing no facility or patients 
image

**Comment by AshrafMd-1** [comment]
```
{"model": "users.user",
        "pk": 23,
        "fields": {
            "password": "argon2$argon2id$v=19$m=102400,t=2,p=8$WDE0YXVzR1dIV09zNWJZalF4QzdXSw$ZTpeDj5sbB3TKfu0DEJ5EkQnL3EmM2vGgvdx+Qt2sAw",
            "last_login": "2024-03-18 17:04:08.753272+00",
            "is_superuser": false,
            "first_name": "Dummy",
            "last_name": "Nurse",
            "email": "dummynurse1@test.com",
            "is_staff": false,
            "is_active": true,
            "date_joined": "2024-03-08 06:31:23.199154+00",
            "username": "dummynurse1",
            "user_type": 14,
            "created_by": 1,
            "ward": null,
            "local_body": null,
            "district": 7,
            "state": 1,
            "phone_number": "+918878825662",
            "alt_phone_number": "+918878825662",
            "gender": 2,
            "age": 21,
            "home_facility": 1,
            "verified": true,
            "deleted": false,
            "pf_endpoint": null,
            "pf_p256dh": null,
            "pf_auth": null,
            "asset": null,
            "groups": [],
            "user_permissions": []
            }
```
            
Also here is the data for nurse 1 is there any wrong?

cc: @rithviknishad

**Comment by AshrafMd-1** [comment]
@nihal467 once this PR https://github.com/coronasafe/care/pull/2012 is merged i will  test it again 

Also I will create a new PR as there are test commits here

**Comment by AshrafMd-1** [comment]
Add this PR has many test commits I will close and open a new PR

**Pull request closed without merging**